proposals/NNNN-package-manager-product-definitions.md

```diff
@@ -7,135 +7,123 @@
+
+To support this, the `dependencies` parameter of the `Target()` initializer will be extended to also allow product references.
+
+To see how this works, remember that each string listed for the `dependencies` parameter is just shorthand for `.target(name: "...")`, i.e. a dependency on a target within the same package.
     
-    This could be extended to support product dependencies in addition to target dependencies.
+For example, the target definition:
     
-    ```
-    let package = Package(
-        name: "Hello",
-        dependencies: [
-            .Package(url: "https://github.com/ExamplePackage", majorVersion: 1),
-        ],
-        targets: [
-            Target(name: "Foo"),
-            Target(name: "Bar", dependencies: [
-                .target(name: "Foo"),
-                .product(name: "Baz", package: "ExamplePackage")
-            ])
-        ],
-    )
-    ```
+```
+Target(name: "Foo", dependencies: ["Bar", "Baz"])
+```
     
-    In this example, the package name is the short name of the package as defined by the package itself.  A possibility would be to allow the package parameter to be optional when it is unambiguous.
+is shorthand for:
     
-    A significant drawback of this approach is the loss of an ability to use strings as unambiguous shorthand for `.target(name: "...")` values.  To remedy this, strings could still be allowed, as long as there was a clear way to disambiguate the dependency in cases in which two or more entities have the same name.
+```
+Target(name: "Foo", dependencies: [.target(name: "Bar"), .target(name: "Baz")])
+```
+
+This will be extended to support product dependencies in addition to target dependencies:
     
-    If we look at practical use cases, it seems rare that two completely unrelated entities (e.g. a target and a product) would have the same name.  For example, if a target depends on `libAlamofire`, it doesn't actually matter whether it's the target or the product that is the dependency.  So in most cases, a string would unambiguously capture the intent of the package author.  If the shorthand form followed a search order, such as:
+```
+let package = Package(
+    name: "MyClientLib",
+    dependencies: [
+        .package(url: "https://github.com/Alamofire/Alamofire", majorVersion: 3),
+    ],
+    targets: [
+        Target(name: "MyUtils"),
+        Target(name: "MyClientLib", dependencies: [
+            .target(name: "MyUtils"),
+            .product(name: "Alamofire", package: "Alamofire")
+        ])
+    ]
+)
+```
     
-    - first look for a target with the specified name in the same package
-    - second, look for a product with the specified name in any package dependency
+The package name is the canonical name of the package, as defined in the manifest of the package that defines the product.  The product name is the name specified in the product definition of that same manifest.
```

Comment:
That was in the earlier proposal but I dropped it from this one because I thought that in the last meeting, we agreed that if we had the string short-form that could bind to either a package or a target, we shouldn't also have the package name be optional here.  I personally would prefer to have several levels of specificity so it isn't all-or-nothing, so I would be happy to add back in optionality of the package name.